src/include/commands/trigger.h

```diff
@@ -0,0 +1,88 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// trigger.h
+//
+// Identification: src/include/commands/trigger.h
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+
+#pragma once
+
+#include "vector"
+#include "expression/abstract_expression.h"
+#include "planner/create_plan.h"
+#include "common/logger.h"
+#include "storage/tuple.h"
+
+namespace peloton {
+namespace commands {
+
+class Trigger {
+ public:
+  Trigger(const planner::CreatePlan& plan);
+  Trigger(const Trigger& that) {
+    trigger_name = that.trigger_name;
+    trigger_funcname = that.trigger_funcname;
+    trigger_args = that.trigger_args;
+    trigger_columns = that.trigger_columns;
+    trigger_when = that.trigger_when->Copy();
+    trigger_type = that.trigger_type;
+  }
+  ~Trigger() {
+    if (trigger_when) {
+      delete trigger_when;
+    }
+  }
+  inline int16_t GetTriggerType() { return trigger_type; }
+  inline std::string GetTriggerName() { return trigger_name; }
+  storage::Tuple* ExecCallTriggerFunc(storage::Tuple *new_tuple);
+
+ private:
+  std::string trigger_name;
+  std::vector<std::string> trigger_funcname;
+  std::vector<std::string> trigger_args;
+  std::vector<std::string> trigger_columns;
+  expression::AbstractExpression* trigger_when = nullptr;
+  int16_t trigger_type;  // information about row, timing, events, access by
+                         // pg_trigger
+};
+
+typedef enum TriggerType {
+  BEFORE_INSERT_ROW = 0,
+  BEFORE_INSERT_STATEMENT,
+  BEFORE_UPDATE_ROW,
+  BEFORE_UPDATE_STATEMENT,
+  BEFORE_DELETE_ROW,
+  BEFORE_DELETE_STATEMENT,
+  AFTER_INSERT_ROW,
+  AFTER_INSERT_STATEMENT,
+  AFTER_UPDATE_ROW,
+  AFTER_UPDATE_STATEMENT,
+  AFTER_DELETE_ROW,
+  AFTER_DELETE_STATEMENT,
+  TRIGGER_TYPE_MAX  // for counting the number of trigger types
+} EnumTriggerType;
+
+class TriggerList {
+ public:
+  inline bool HasTriggerType(EnumTriggerType type) const {
+    return types_summary[type];
+  }
+  inline int GetTriggerListSize() { return static_cast<int>(triggers.size()); }
+  void AddTrigger(Trigger trigger);
+  void UpdateTypeSummary(int16_t type);
+  Trigger* Get(int n) { return &triggers[n]; }  // get trigger by index
```

Comment:
Thanks. Yes, this should be more defensive. Actually, I don't think it reasonable to expose a `Get` interface here. This function might be created by a team member for debugging. I'll try to remove this in the future.